backend/main.py

```diff
@@ -2,70 +2,86 @@
-    '''Desample power data
-
-        Read power data from given filename, desample and limit number of records to the given number.
-        The desample strategy is to select one data point per certain number to ensure that output 
-        data points are kept under threshold.
-
-        Args:
-            file_name: The filename of data csv
-            num_records: Total number of data points 
-            max_records: Max number of data points to output
-        
-        Returns:
-            A list of power data. For example:
-            [
-                ['1532523212', '53', 'SYSTEM'],
-                ['1532523242', '33', 'SYSTEM']
-            ]
-    '''
-    data = list()
-    frequency = num_records / max_records
-
-    with open(file_name, 'r') as fr:
-        for i, line in enumerate(fr):
-            if i % frequency == 0:
-                data.append(line.strip('\n').split(','))
-    return data
-    
+@app.route('/data')
+def get_data():
+    """HTTP endpoint to get data
 
+        Retrives all power data from local file given a limit on number of
+        records from request body.
+    """
+    strategy = request.args.get('strategy', default='avg', type=str)
+    start = request.args.get('start', default=None, type=int)
+    end = request.args.get('end', default=None, type=int)
+    if not strategy in STRATEGIES:
+        return 'Incorrect Strategy', 400
+
+    cache_filename = utils.cache_filename(FILENAME, strategy)
+    data = downsample.secondary_downsample(
+        cache_filename, strategy, NUMBER_OF_RECORDS_PER_REQUEST, start, end)
 
-@app.route('/data')
-def getData():
-    '''HTTP endpoint to get data
-        
-        Retrives all power data from local file given a limit on number of records from request body.
-    '''
-
-    fn = 'Power_sample_data.csv'
-    num_records = 7200000  # assume number of records is accessible upon deployment
-    max_records = request.args.get('number', default=6000, type=int)
-    
-    data = downsample_raw_data(fn, num_records, max_records)
-    # TODO(tangyifei@): Implementation too naive, will improve with selected time interval, better desample strategy.
     return jsonify(data)
 
 
+@app.route('/preprocessing')
+def preprocessing():
+    """HTTP endpoint to preprocess data
+
+        Preprocess data for all strategy and save results locally,
+        each strategy in a file.
+    """
+    for strategy in STRATEGIES:
+        data = downsample.downsample(
+            FILENAME, strategy, NUMBER_OF_RECORDS_PER_SECOND)
+        data_csv = utils.csv_format(data)
+        output_filename = utils.cache_filename(FILENAME, strategy)
+        with open(output_filename, 'w') as filewriter:
+            filewriter.write(data_csv)
+            filewriter.flush()
+    return 'Preprocessing Successful!'
+
+
+@app.before_first_request
+def initialize():
+    """Initialize with preprocessing
+
+        Preprocess all power data at the first request.
```

Comment:
The power data will be continuously uploaded to a place. So, better start a cron job to find all the newly added files and then preprocess them.

Reply:
Will support finding new files as we get to gcp bucket stuff.